opencog/query/Implicator.h

```diff
@@ -47,20 +47,31 @@ namespace opencog {
  *
  * The 'var_soln' argument in the callback contains the map from variables
  * to ground terms. 'class Instantiator' is used to perform the actual
- * grounding.  A list of grounded expressions is created in 'result_list'.
+ * grounding.  A set of grounded expressions is created in 'result_set'.
+ * Note that the callback may be called many times reporting the same
+ * results. In that case the 'result_set' will contain unique solutions.
  */
 class Implicator :
 	public virtual PatternMatchCallback
 {
+	protected:
+		UnorderedHandleSet _result_set;
+		HandleSeq _result_list;
```

Comment:
So  get_result_list() returns a HandleSeq; that HandleSeq will be constructed on-stack, and returned to the caller. There is no point in caching _result_list: you are just forcing a copy from the HandleSeq instance in this class instance, to the callers stack frame. 

I guess you are thinking that, byt using the _result_changed flag, you are saving some cpu time. if the user asks for the result_list more than once ... but I don't see that being likely?